### PR TITLE
Noval

### PR DESCRIPTION
**Pull request by ZacLN**
Some breaking changes:
- leaf node types are no long `EXPR`
- only IDENTIFIER and LITERAL store a `.val` 
- operator calls have fixed number of args (I may revert this, will do some speed tests)
- remove precedence parameter on `OPERATOR`

@keno , as the other user of this, thoughts?

### CODE REVIEW

**Comment by Keno** [comment]
Looks fine to me.

**Comment by KristofferC** [comment]
FWIW, I am doing some local work on Tokenize to 1) not allocate new strings for each "deterministic" token and 2) support using completely `val`-less tokens. Unfortunately, the coming two weeks I am pretty dead on time so it will be a while.

**Comment by ZacLN** [comment]
Great, I fiddled a bit with that and it seemed like it could give a 2-3x speed up?

**Comment by KristofferC** [comment]
Out of curiosity, any performance improvements?

**Comment by ZacLN** [comment]
with the tokenize changes it's about 2x

**Comment by KristofferC** [comment]
Cool :)

**Comment by ZacLN** [comment]
The main problem is trying to precompile it properly, the first run of the speed test (repeatedly parsing some big files from `base`) takes ~25 seconds then about 0.8s afterwards

**Comment by KristofferC** [comment]
Hmm, the first run for collecting all tokens in a large file with Tokenize takes around 0.5 seconds so the overhead there seems to not represent the 25 seconds.

**Comment by ZacLN** [comment]
Yeah, it's just the first bit of inference it hits [PR](https://github.com/KristofferC/Tokenize.jl/pull/112)

File: src/utils.jl [hunk 1]
```diff
@@ -179,11 +179,13 @@ ispunctuation(t::Token) = t.kind == Tokens.COMMA ||
 
 isstring(x) = false
 isstring(x::EXPR{StringH}) = true
-isstring(x::LITERAL{Tokens.STRING}) = true
-isstring(x::LITERAL{Tokens.TRIPLE_STRING}) = true
+isstring(x::LITERAL) = x.kind == Tokens.STRING || x.kind == Tokens.TRIPLE_STRING
+is_integer(x) = x isa LITERAL && x.kind == Tokens.INTEGER
+is_float(x) = x isa LITERAL && x.kind == Tokens.FLOAT
+is_number(x) = x isa LITERAL && (x.kind == Tokens.INTEGER || x.kind == Tokens.FLOAT)
+is_nothing(x) = x isa LITERAL && x.kind == Tokens.begin_plus
```
Comment (KristofferC) [inline src/utils.jl]: Huh?

Comment (ZacLN) [reply src/utils.jl]: I need to fix this, I use a couple barrier tokens as Token kinds in CSTParser. I'll add a PR which does so and removes the `invisible_keyword`s from Tokenize

File: src/utils.jl [hunk 2]
```diff
@@ -83,61 +83,81 @@ macro precedence(ps, prec, body)
     end
 end
 
+struct TMP
```
Comment (KristofferC) [inline src/utils.jl]: I named these guys kinda bad so feel free to rename them. A comment of why it's used (to reduce generated) code might be useful as well.

**Pull request merged**